### ISSUE TITLE
[Impeller] Reduce render-pass overhead when using wide gamut (iOS)

### ISSUE THREAD

**Issue by knopp** [flutter/flutter#131567]
Consider the following example:
<details>
 <summary>code</summary>

```dart
import 'dart:ui';
import 'package:flutter/material.dart';

void main() {
  runApp(const MainApp());
}

const kSigma = 1.0;
const kNumberOfBlurs = 6;

class _Blur extends StatelessWidget {
  const _Blur();

  @override
  Widget build(BuildContext context) {
    return ClipRRect(
      child: BackdropFilter(
        blendMode: BlendMode.srcIn,
        filter: ImageFilter.blur(
          sigmaX: kSigma,
          sigmaY: kSigma,
        ),
        child: Container(
          color: Colors.red.withAlpha(30),
          child: const Text('Blur'),
        ),
      ),
    );
  }
}

class MainApp extends StatelessWidget {
  const MainApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Stack(
          children: [
            ListView.builder(
                itemBuilder: (context, index) {
                  return Container(
                    padding: const EdgeInsets.all(10),
                    child: Text('Item $index'),
                  );
                },
                itemCount: 1000),
            for (var i = 0; i < kNumberOfBlurs; ++i) ...[
              Positioned(
                left: 0,
                right: 0,
                top: i * 150,
                height: 60,
                child: const _Blur(),
              ),
            ]
          ],
        ),
      ),
    );
  }
}
```
</details>

<details>
<summary>screenshot</summary>

![Screen Shot 2023-07-30 at 09 49 18](https://github.com/flutter/flutter/assets/96958/4ae73552-8960-4778-81a0-7ef5c59e5c4a)
</details>

It has 6 blur backdrops with sigma=1 (to focus on render pass overhead and not blur filter performance). 

This is the performance on A15 (iPhone 13 Pro) with wide gamut disabled:
<img width="540" alt="Screenshot 2023-07-29 at 16 40 56" src="https://github.com/flutter/flutter/assets/96958/6d6426df-f53d-41d2-90df-6bbadc9e06db">

and wide gamut enabled:
<img width="541" alt="Screenshot 2023-07-29 at 16 42 13" src="https://github.com/flutter/flutter/assets/96958/8f20a94e-d6c0-4cee-8e23-9d9e724df080">

I tried couple of things to figure out why the regression is so large:

I noticed that there is quite a lot of texture allocation per frame. Caching textures improved the frame time by around 1-1.5ms.

Biggest bottleneck by far seems to be the blits for `MSAA backdrop`. They take around 80% of frame time.

Disabling `MSAA backdrop` (reverting back to `LoadAction::kLoad` + `StoreAction::kStoreAndMultisampleResolve`/`kMultisampleResolve`) seems to improve the performance considerably:
<img width="382" alt="Screenshot 2023-07-29 at 22 42 22" src="https://github.com/flutter/flutter/assets/96958/fdecb74e-6e3e-44a7-8716-155248d29c54">

cc @gaaclarke, @bdero, @jonahwilliams

**Comment by knopp** [comment]
> > It has 6 blur backdrops with sigma=1 (to focus on render pass overhead and not blur filter performance).
> 
> The amount of work required to calculate the blurs doesn't depend on the sigma value. Check out `IPGaussian` in `gaussian.glsl`.

The sigma (as specified in `ImageFilter.blur`) is used to calculate blur radius, which affects the number of texture samples inside `gaussian_blur.glsl`. The number of of image samples seems to be the biggest bottleneck currently (see https://github.com/flutter/flutter/issues/131573). What am I missing here?

**Comment by gaaclarke** [comment]
Thanks, that makes more sense.  I didn't know we don't have a separate knob for radius at the dart level.

**Comment by chinmaygarde** [comment]
This is related to making all blurs faster and just happens to show up more urgently when wide-gamut is enabled. That work is tracked in https://github.com/flutter/flutter/issues/131580. Closing as duplicate.

**Comment by knopp** [comment]
@chinmaygarde, the `MSAA backdrop` blit being significantly slower than `LoadAction::kLoad + StoreAction::kStoreAndMultisampleResolve` is an issue regardless of blur. `MSAA blackdrop` was introduced as an optimisation, but that's not always the case, shouldn't this be tracked?

**Comment by chinmaygarde** [comment]
Ah, my bad. That makes sense. Can we reword this to describe the action-item instead of the symptom?

**Comment by jonahwilliams** [comment]
@knopp I'm not getting similar numbers to you, can you post your full trace for this application?

**Comment by knopp** [comment]
Attached are traces for current main with wide gamut enabled and disabled. Tomorrow I'll try to publish a branch with MSAA backdrop disabled and appropriate. This is on iPhone 13 Pro.

[Traces.zip](https://github.com/flutter/flutter/files/12339000/Traces.zip)

**Comment by knopp** [comment]
@jonahwilliams, I was playing with this and it seems initially I missed that for MSAA backdrop removal to perform better, it is also necessary to reuse textures. Without reusing textures removing MSAA backdrop performs much worse.

[no-msaa-backdrop-reuse-textures dart_devtools_2023-08-15_01_11_41.977.json.zip](https://github.com/flutter/flutter/files/12340139/no-msaa-backdrop-reuse-textures.dart_devtools_2023-08-15_01_11_41.977.json.zip)

Here is my branch I tested this with: https://github.com/knopp/engine/commits/remove_msaa_backdrop
It's very hacky, but it can render the example sub 8ms on iPhone 13 pro, compared to ~12ms on main.

**Comment by jonahwilliams** [comment]
FWIW we've started recylcing render target textures after https://github.com/flutter/engine/pull/44527

Very simple herustics, but seem to get a good hit rate locally.
